### PR TITLE
tests, utils: Drop NewRandomVMWithDataVolumeAndUserDataInStorageClass

### PR DESCRIPTION
**Pull request by akalenyu**
### What this PR does
Dropping NewRandomVMWithDataVolumeAndUserDataInStorageClass for direct libvmi usage.

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
NONE
```

### CODE REVIEW

**Review by dankenigsberg — CHANGES_REQUESTED** [review]
thank you for making our code better. please consider using simpler static function definition.

**Comment by akalenyu** [comment]
After some experimenting with a factory that creates a persistent-disk VM I am not sure it has any benefits over direct libvmi/libdv calls. Hiding DV/VMI spec under such an API feels wrong, and exposing it fully brings us back to the first step.

I am not against making the shortcut function static, but that would basically now mean it's an API under the storage package.
Is this something we want @dankenigsberg? I think it's purely lucky that this definition works for both suites

**Comment by dankenigsberg** [comment]
> I am not against making the shortcut function static, but that would basically now mean it's an API under the storage package.

I don't think I understand your question. You have a `storage` package with several files in it. It makes a lot of sense to me that `export` and `restore` would creating similar kinds of VMs. It feels natural to place it in a `vmifactory.go` file next to them. You may even want to expose it for usage by other test packages (migration?).

Using a function variable for something that never varies is a bad practice. It adds mental burden (you keep worrying that it may change), it mixes implementation and usage, it makes the calling context longer and more complicated, it often captures unintended variable. You wouldn't do it production code, so you shouldn't do it in test code, either.

**Comment by akalenyu** [comment]
> It makes a lot of sense to me that `export` and `restore` would creating similar kinds of VMs

This is the point I was referring to - this may be true now, and particularly for these existing usages,
but may not be in the future when somebody picks it up for other storage tests

**Comment by dankenigsberg** [comment]
> This is the point I was referring to - this may be true now, and particularly for these existing usages,
but may not be in the future when somebody picks it up for other storage tests

The function you create should have a very clear name that tells what VMI it is generating. Future users would have to make a call if they reuse it, or create their own factory function. As long as you keep it private to the `storage` package you have a high level of flexibility to change it or its name in the future - it's not yet a public API.

File: tests/storage/restore.go [hunk 1]
```diff
@@ -1868,3 +1837,26 @@ var _ = SIGDescribe("VirtualMachineRestore Tests", func() {
 		})
 	})
 })
+
+func newVMWithDataVolumeForRestore(storageClass string) *v1.VirtualMachine {
```
Comment (dankenigsberg) [inline tests/storage/restore.go:1841]: I would prefer one reusable function, but this can wait for the third user.

**Review by dankenigsberg — APPROVED** [review]

**Comment by iholder101** [comment]
Thanks for this PR @akalenyu!

> I don't think I understand your question. You have a storage package with several files in it. It makes a lot of sense to me that export and restore would creating similar kinds of VMs. It feels natural to place it in a vmifactory.go file next to them. You may even want to expose it for usage by other test packages (migration?).

I 100% agree.
If the function name reflects which VM is being created, what's the motivation of duplicating the function over different files? I think that a `vmifactory.go` file under the storage test package is definitely the right direction. As @dankenigsberg said, it would both be helpful for other storage test packages, but perhaps would also be valuable for different tests.

I would suggest:
1) Give this function a good name explaining exactly which VMI it's generating
2) Get options as parameters so others could easily make small modifications to the VMI being created.

I imagine something like:
```go
func GreatNameForExportDVCreation(extraDvOpts... []libdv.Option) {...}
func GreatNameForExportVmiCreationWithDV(dv *DataVolume, extraVmiOpts... libvmi.Option) {...}
func GreatNameForExportVmiCreation(extraVmiOpts... libvmi.Option) {...}
```

WDYT @akalenyu?

**Comment by akalenyu** [comment]
> Thanks for this PR @akalenyu!
> 
> > I don't think I understand your question. You have a storage package with several files in it. It makes a lot of sense to me that export and restore would creating similar kinds of VMs. It feels natural to place it in a vmifactory.go file next to them. You may even want to expose it for usage by other test packages (migration?).
> 
> I 100% agree. If the function name reflects which VM is being created, what's the motivation of duplicating the function over different files? I think that a `vmifactory.go` file under the storage test package is definitely the right direction. As @dankenigsberg said, it would both be helpful for other storage test packages, but perhaps would also be valuable for different tests.
> 
> I would suggest:
> 
> 1. Give this function a good name explaining exactly which VMI it's generating
> 2. Get options as parameters so others could easily make small modifications to the VMI being created.
> 
> I imagine something like:
> 
> ```go
> func GreatNameForExportDVCreation(extraDvOpts... []libdv.Option) {...}
> func GreatNameForExportVmiCreationWithDV(dv *DataVolume, extraVmiOpts... libvmi.Option) {...}
> func GreatNameForExportVmiCreation(extraVmiOpts... libvmi.Option) {...}
> ```
> 
> WDYT @akalenyu?

I explored similar ideas (https://github.com/kubevirt/kubevirt/pull/11476#issuecomment-2004446098).
What I found is that you end up with an inferior variation of the existing factories (limited by what is customizable, for instance, no VMOptions);

Ideally, we could just use the factories directly, it's just that this becomes too lengthy.
Correct me if I'm wrong but we end up with the same bloat if we proceed with this.

**Comment by iholder101** [comment]
> I explored similar ideas ([#11476 (comment)](https://github.com/kubevirt/kubevirt/pull/11476#issuecomment-2004446098)). What I found is that you end up with an inferior variation of the existing factories (limited by what is customizable, for instance, no VMOptions);
>
> Ideally, we could just use the factories directly, it's just that this becomes too lengthy. Correct me if I'm wrong but we end up with the same bloat if we proceed with this.

I wouldn't say that.
Our current factories are very generic and generate very basic VMIs. Something more complex as a vmi suitable for export is something that requires both knowledge and bloat.

If someone who's not a storage expert would try to create such a VMI with libvmi only it would be completely non-intuitive.  On the other hand, someone writing a migration test might need to use such a VMI. How would you suggest someone like that create such a VMI, by copy-pasting the function you made for the third time now? This would create much more bloat and mess than simply letting others re-use the VMI you've constructed.

By creating one exported function, with `(extraVmiOpts... libvmi.Option)` as parameter, you both let others re-use this VMI and give them an opportunity to add minor modifications to it.

Does it make sense to you?

Either way I don't think it should block this PR. If you feel strongly against it, you're welcome to remove the hold. Thanks for your contribution!

/approve
/hold

**Comment by dankenigsberg** [comment]
I'd prefer more code reuse in this case, but I prefer even more to see this PR merged over the weekend

/hold cancel

**Pull request merged**